### PR TITLE
feat(config): add support for Namron Dimmer 2 400W

### PR DESCRIPTION
**Pull request by GD-Dal**
First time pull request, so please don't shoot me.
Trying to create support for a Namron dimmer.
I've created a js file, where it's information is taken directly from the manual that is linked below.

Here is a link to the product:
https://www.namron.com/products/namron-z-wave-dimmer-2-400w/
And here is a link to the manual:
https://www.elektroimportoren.no/docs/lib/4512724-Brukerveiledning-5.pdf

### CODE REVIEW

File: namron_dimmer.js [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+		{
+			"productType": "0x0200",
+			"productId": "0xd00c"
+		}
+	],
+	"firmwareVersion": {
+		"min": "0.0",
+		"max": "255.255"
+	},
+	"supportsZWavePlus": true,
```
Comment (AlCalzone) [inline namron_dimmer.js]:
```suggestion
```
We do not use that anymore.

**Review by AlCalzone — CHANGES_REQUESTED** [review]
Probably best if you work through the [style guide](https://zwave-js.github.io/node-zwave-js/#/config-files/style-guide) before we do an in-depth review. Especially the labels and descriptions need to be cleaned up - and multiline strings are not supported in JSON.

**Review by GD-Dal — COMMENTED** [review]
removed "supportsZWavePlus
changed all descriptions to be on one line only.

File: namron_dimmer.js [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+			  },
+			  {
+					 "label": "Enable to send Basic report",
+					 "value": 1
+			  }
+		   ]
+		},
+		"4": {
+			"label": "Default fade time (in seconds)",
+			"description": "Default fade time (unit is second, this value has the same function as Duration of Multilevel) Valid value: 0~127",
```
Comment (AlCalzone) [inline namron_dimmer.js]: Quote from the style guide:
> unnecessary descriptions that merely restate the label must be removed. Additionally, information like units or available ranges should be removed as that information is provided to UIs through other properties.

**Review by AlCalzone — CHANGES_REQUESTED** [review]
Same goes for the others. Like I said - please read the style guide carefully and apply it.

File: namron_dimmer.js (hunk 2 above)
Comment (GD-Dal) [reply namron_dimmer.js]: Ok, thanks for guiding me.
I've edited the file now. Hopefully it looks better now.

**Comment by blhoward2** [comment]
Also, this file is misnamed. It should end in .json

**Comment by GD-Dal** [comment]
> 
> 
> Also, this file is misnamed. It should end in .json

Ok, changed it now

**Comment by AlCalzone** [comment]
There's several things wrong, most of which should be apparent when you read the documentation

1. the file is in the wrong location => https://zwave-js.github.io/node-zwave-js/#/config-files/overview
2. the file has syntax errors, missing commas, etc.
3. At least the labels and descriptions don't follow the [style guide](https://zwave-js.github.io/node-zwave-js/#/config-files/style-guide) that I've linked you above.

> Parameter descriptions can be helpful, but they clutter UIs. As such, unnecessary descriptions that merely restate the label must be removed. Additionally, information like units or available ranges should be removed as that information is provided to UIs through other properties.

File: namron_dimmer.json [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+		"2": {
+			"label": "Save load state",
+			"description": "Saves load state before power failure",
+			"valueSize": 1,
+			"minValue": 0,
+			"maxValue": 2,
+			"defaultValue": 2,
+			"readOnly": false,
+			"writeOnly": false,
+			"allowManualEntry": true
```
Comment (AlCalzone) [inline namron_dimmer.json]: missing ,

File: namron_dimmer.json [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+			  {
+					 "label": "Shut off load",
+					 "value": 0
+			  },
+			  {
+					 "label": "Turn on load",
+					 "value": 1
+			  },
+			  {
+					 "label": "Save load state before power failure"
```
Comment (AlCalzone) [inline namron_dimmer.json]: missing ,

File: namron_dimmer.json [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+		"3": {
+			"label": "Enable/disable to send report",
+			"description": " Sends the basic report to the Lifeline when the load state changed（When value set as 1, repower on the dimmer, it will send Basic report automatically）",
+			"valueSize": 1,
+			"minValue": 0,
+			"maxValue": 1,
+			"defaultValue": 1,
+			"readOnly": false,
+			"writeOnly": false,
+			"allowManualEntry": true
```
Comment (AlCalzone) [inline namron_dimmer.json]: missing ,

File: namron_dimmer.json [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+			  },
+			  {
+					 "label": "Save load state before power failure"
+					 "value": 2
+			  }
+		   ]	 
+		},
+		"3": {
+			"label": "Enable/disable to send report",
+			"description": " Sends the basic report to the Lifeline when the load state changed（When value set as 1, repower on the dimmer, it will send Basic report automatically）",
```
Comment (AlCalzone) [inline namron_dimmer.json]: Very long, has leading space

File: namron_dimmer.json [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+			"minValue": 0,
+			"maxValue": 50,
+			"defaultValue": 15,
+			"readOnly": false,
+			"writeOnly": false,
+			"allowManualEntry": true
+		},
+		"6": {
+			"label": "Maxium brightness value",
+			"description": "Valid values: 0 ~ 100",
```
Comment (AlCalzone) [inline namron_dimmer.json]: Style guide:
> information like units or **available ranges should be removed** as that information is provided to UIs through other properties.

File: namron_dimmer.json [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+			"productId": "0xd00c"
+		}
+	],
+	"firmwareVersion": {
+		"min": "0.0",
+		"max": "255.255"
+	},
+	"paramInformation": {
+		"2": {
+			"label": "Save load state",
```
Comment (AlCalzone) [inline namron_dimmer.json]: Style guide:
> Labels should be **Title Case**.

**Review by AlCalzone — CHANGES_REQUESTED** [review]
I've linked you some examples - these are not all. Please go through the style guide carefully and apply it.

File: namron_dimmer.json (hunk 7 above)
Comment (GD-Dal) [reply namron_dimmer.json]: Hi, and thanks for the input.
I have gone through the JSON code again.
After reading the Style Guide, it is my understanding that the parameters readOnly, writeOnly are not needed unless the manual states that the value is in fact Read Only?
And the allowManualEntry must be TRUE where the user can set a VALUE between given values, and must be set to FALSE where unchangable OPTIONS are set
And: What is the "unsigned": true for? I see it several places, but the Style Guide does not explain what it is.

I have also run the code through an online JSONLint checker, and it detects no errors now.
So, do you mind looking over the code once again?

Thank you

**Comment by GD-Dal** [comment]
Hi, and thanks for the input.
I have gone through the JSON code again.
After reading the Style Guide, it is my understanding that the parameters readOnly, writeOnly are not needed unless the manual states that the value is in fact Read Only?
And the allowManualEntry must be TRUE where the user can set a VALUE between given values, and must be set to FALSE where unchangable OPTIONS are set
And: What is the "unsigned": true for? I see it several places, but the Style Guide does not explain what it is.

I have also run the code through an online JSONLint checker, and it detects no errors now.
So, do you mind looking over the code once again?

About the file placement: Is it possible to move the existing file, or do I need to go in there and create a new file. And then paste in the contents of the old one?

Thank you

**Comment by AlCalzone** [comment]
> After reading the Style Guide, it is my understanding that the parameters readOnly, writeOnly are not needed unless the manual states that the value is in fact Read Only?

correct

> And the allowManualEntry must be TRUE where the user can set a VALUE between given values, and must be set to FALSE where unchangable OPTIONS are set

Correct, but `true` is the default for writeable parameters. So you only need to specify this and set it to false, if you want to limit the entry to the predefined options.

> Is it possible to move the existing file, or do I need to go in there and create a new file. And then paste in the contents of the old one?

Not sure if this is possible through the Github UI, but I can do that for you locally.

> What is the "unsigned": true for?

Per the specifications, Z-Wave config parameters are signed, so for a `valueSize` of `1`, you'd have a value range of -128..127. Many devices interpret their parameters as unsigned, which corresponds to a value range of 0..255 for a `valueSize` of `1`. There's no hard rule for when to use `unsigned`, except when the min/max values aren't compatible with the signed range. If that's the case and you didn't notice, the automatic checks will catch this and warn you.

**Comment by AlCalzone** [comment]
@GD-Dal I've done a few more fixes - mainly removed `allowManualEntry` where it wasn't allowed. Also, some labels needed cleanup for consistency and there were a few more unnecessary descriptions: [`845be02` (#2528)](https://github.com/zwave-js/node-zwave-js/pull/2528/commits/845be02657afc554ee4a1347bc074d78cad41381)
Please keep these changes in mind for the next time :)

**Review by AlCalzone — DISMISSED** [review]
Approved with my changes

**Comment by GD-Dal** [comment]
> Please keep these changes in mind for the next time :)

I will :)
Thank you for all your help.
What is the next step now? Is there anything I need to do?
How does it take until it's actually in the z-wave js database?

Thanks again

**Comment by AlCalzone** [comment]
I'll have to merge the PR, then it will take a couple of days until I release the next version. The software you use needs to be updated too, so that will take another couple of days.
In the meantime, if you use zwavejs2mqtt you should be able to copy the final file into the custom config directory and start using it. I'm not sure where that option is, @robertsLando should know.

**Comment by robertsLando** [comment]
@GD-Dal the directory for custom devices config is 

`deviceConfigPriorityDir: storeDir + '/config'`

**Pull request merged**